File: algorithms/2_medium/R/00519_random_flip_matrix/vector.c
typedef struct {
    int valids_len;
    int m;
    int n;
    int valids[];
} Solution;

Solution* solutionCreate(int m, int n) {
    assert(n >= 1 && m >= 1 && n <= 10000 && m <= 10000);
    const int total = n * m;
    Solution* res = malloc(sizeof(*res) + (size_t)total * sizeof(int));
    if (res == NULL) { return NULL; }

    for (int i = 0; i < total; i += 1) {
        res->valids[i] = i;
    }
    res->valids_len = total;
    res->n = n;
    res->m = m;

    return res;
}

int* solutionFlip(Solution* obj, int* return_len) {
    assert(return_len != NULL && obj != NULL);
    if (obj->valids_len == 0) {
        *return_len = 0;
        return NULL;
    }

    int *ret_vec = malloc(2 * sizeof(*ret_vec));
    if (ret_vec == NULL) {
        *return_len = 0;
        return NULL;
    }

    const int random_value = (int)(random() % obj->valids_len);
    const int res = obj->valids[random_value];
    obj->valids[random_value] = obj->valids[obj->valids_len - 1];
    obj->valids[obj->valids_len - 1] = res;

    obj->valids_len -= 1;
    *return_len = 2;
    ret_vec[0] = res / obj->n;
    ret_vec[1] = res % obj->n;

    return ret_vec;
}

void solutionReset(Solution* obj) {
    const int total = obj->n * obj->m;
    obj->valids_len = total;
}

void solutionFree(Solution* obj) {
    free(obj);
}

/**
 * Your Solution struct will be instantiated and called as such:
 * Solution* obj = solutionCreate(m, n);
 * int* param_1 = solutionFlip(obj, retSize);

 * solutionReset(obj);

 * solutionFree(obj);
*/

File: algorithms/2_medium/R/00519_random_flip_matrix/vector.rs
struct Solution {
    valids: Vec<i32>,
    valids_len: usize,
    m: i32,
    n: i32,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl Solution {

    fn new(m: i32, n: i32) -> Self {
        assert!(n >= 1 && m >= 1 && n <= 10_000 && m <= 10_000);
        let total = n * m;

        Self {
            valids: (0..total).collect(),
            valids_len: total as usize,
            m: m,
            n: n,
        }
    }

    fn flip(&mut self) -> Vec<i32> {
        if self.valids_len == 0 {
            return Vec::default();
        }

        use rand::Rng;
        let mut rng = rand::thread_rng();
        let random_value: usize = rng.gen_range(0..self.valids_len);
        let res = self.valids[random_value];
        self.valids[random_value] = self.valids[self.valids_len - 1];
        self.valids[self.valids_len - 1] = res;

        self.valids_len -= 1;

        vec![res / self.n, res % self.n]
    }

    fn reset(&mut self) {
        let total = self.n * self.m;
        self.valids_len = total as usize;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * let obj = Solution::new(m, n);
 * let ret_1: Vec<i32> = obj.flip();
 * obj.reset();
 */
